docs/design/module_design.md

```diff
@@ -0,0 +1,127 @@
+# Module Design
+## Namespace
+Namespace is an isolation concept in Satellite. Each namespace has one agent to collect the APM data.
+So many agents would be running at the same time in Satellite.
+```
+                            Satellite
+ ---------------------------------------------------------------------
+|            -------------------------------------------              |
+|           |                 Agent                     |             |
+|            -------------------------------------------              |
+|            -------------------------------------------              |
+|           |                 Agent                     |             |
+|            -------------------------------------------              |
+|            -------------------------------------------              |
+|           |                 Agent                     |             |
+|            -------------------------------------------              |
+ ---------------------------------------------------------------------
+```
+## Modules
+There are 3 modules in one agent, which are Gatherer, Processor, and Sender.
+
+- The Gatherer module is responsible for fetching or receiving data and pushing the data to Queue. So there are 2 kinds of Gatherer, which are ReceiverGatherer and FetcherGatherer.
+- The Processor module is responsible for reading data from the queue and processing data by a series of filter chains.
+- The Sender module is responsible for async processing and forwarding the data to the external services in the batch mode. After sending success, Sender would also acknowledge the offset of Queue in Gatherer.
+
+```
+                            Agent
+ --------------------------------------------------------------------
+|            ----------      -----------      --------               |
+|           | Gatherer | => | Processor | => | Sender |              |                          
+|            ----------      -----------      --------               |
+ --------------------------------------------------------------------
+```
+
+## Plugins
+
+Plugin is the minimal components in the module. There are 2 kinds of plugins in Satellite, which are sharing plugins and normal plugins.
+
+- a sharing plugin instance could be sharing with multiple modules in the different namespaces.
+- a normal plugin instance is only be used in a fixed module of the fixed namespaces.
+
+### Sharing plugin
+Nowadays, there are 2 sharing plugins in Satellite, which are server plugins and client plugins. The reason why they are sharing plugins is to reduce the resource cost in connection. Server plugins are sharing with the ReceiverGatherer modules in the different namespaces to receive the external requests. And the client plugins is sharing with the Sender modules in the different namespaces to connect with external services, such as Kafka and OAP.
```

Comment:
I think this is not `2 sharding plugins`. You mean there are 2 kinds of sharding plugins?